crates/mun_abi/src/type_lut.rs

```diff
@@ -1,36 +1,81 @@
-use crate::Guid;
-use itertools::izip;
 use std::ffi::CStr;
+use std::hash::Hash;
 use std::os::raw::c_char;
 use std::{ffi, fmt, slice, str};
 
-/// Represents a unique identifier for types. The runtime can use this to lookup the corresponding [`TypeInfo`].
-#[repr(C)]
+use itertools::izip;
+
+use crate::Guid;
+
+/// Represents a unique identifier for types. The runtime can use this to lookup the corresponding
+/// [`TypeInfo`]. A [`TypeId`] is a key for a [`TypeInfo`].
+///
+/// A [`TypeId`] only contains enough information to query the runtime for a concrete type.
+#[repr(u8)]
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub struct TypeId {
-    /// The GUID of the type
-    pub guid: Guid,
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+pub enum TypeId<'a> {
+    /// Represents a concrete type with a specific Guid
+    Concrete(Guid),
+
+    /// Represents a pointer to a type
+    Pointer(PointerTypeId<'a>),
+}
+
+/// Represents a pointer to another type.
+#[repr(C)]
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+pub struct PointerTypeId<'a> {
+    /// The type to which this pointer points
+    pub pointee: &'a TypeId<'a>,
```

Comment:
Should we create an issue to replace all occurrences of `*const` and `*mut` in `mun_abi` with their reference counterparts?

Reply:
Yes, you were going to create an issue for that right?